src/commands/azureServiceOperators/ui/azureservicebrowser.ts

```diff
@@ -16,21 +21,34 @@ export async function AzureServiceBrowser(explorer: k8s.ClusterExplorerV1): Prom
 }
 
 async function allServiceKinds(): Promise<AzureServiceKind[] | undefined> {
+    const apiResult = await getAPIResourceCommandResult();
+
+    if (apiResult.succeeded) {
+        return apiResult.result;
+    } else if (apiResult.error !== ASOInstallation.ASONotInstalled) {
+        vscode.window.showWarningMessage(apiResult.error);
+    }
+    return undefined;
+}
+
+async function getAPIResourceCommandResult(): Promise<Errorable<AzureServiceKind[]>> {
     const kubectl = await k8s.extension.kubectl.v1;
 
     if (!kubectl.available) {
-        vscode.window.showWarningMessage(`Kubectl is unavailable.`);
-        return undefined;
+        return { succeeded: false, error: `Kubectl is unavailable.` };
     }
     const asoAPIResourceCommandResult = await kubectl.api.invokeCommand("api-resources --api-group azure.microsoft.com --no-headers");
-    if (!asoAPIResourceCommandResult || (asoAPIResourceCommandResult.code !== 0 && !asoAPIResourceCommandResult.stdout)) {
-        vscode.window.showWarningMessage(`Azure Service Operator api-resources command failed with following error: ${asoAPIResourceCommandResult?.stderr}.`);
-        return undefined;
-    }
 
-    const treeResourceItems = asoAPIResourceCommandResult.stdout.split("\n").map((line: string) => line.trim()).filter((line: string | any[]) => line.length > 0);
-
-    return treeResourceItems.map((item: string) => parseServiceResource(item)!);
+    if (!asoAPIResourceCommandResult) { // Fail to invoke command.
+        return { succeeded: false, error: `Azure Service Operator api-resources failed to invoke command.` };
+    } else if (asoAPIResourceCommandResult.code !== 0 && !asoAPIResourceCommandResult.stdout) { // Error result and Faild execution.
```

Comment:
```suggestion
    } else if (asoAPIResourceCommandResult.code !== 0 && !asoAPIResourceCommandResult.stdout) { // kubectl errored and there is no partial result
```

Another way to approach this that might be clearer is to say: if there is _any_ stdout then (regardless of exit code) parse and return.  Then tackle the other cases:

```
if (execFailed) { return genericerr }
else if (thereIsStdout) { return success }
else if (code === 0) { return specialNotInstalledError }
else { return stderr }
```

This avoids the complex and slightly counterintuitive if-clause here.